File: packages/modules/lodash/kotlin/src/math.kt
package lodash

import kotlin.math.*

/**
 * Adds two numbers.
 */
fun add(augend: Double, addend: Double): Double = augend + addend
fun add(augend: Int, addend: Int): Int = augend + addend

/**
 * Computes number rounded up to precision.
 */
fun ceil(number: Double, precision: Int = 0): Double {
    val factor = 10.0.pow(precision)
    return ceil(number * factor) / factor
}

/**
 * Divide two numbers.
 */
fun divide(dividend: Double, divisor: Double): Double = dividend / divisor

/**
 * Computes number rounded down to precision.
 */
fun floor(number: Double, precision: Int = 0): Double {
    val factor = 10.0.pow(precision)
    return floor(number * factor) / factor
}

/**
 * Computes the maximum value of array. If array is empty or falsey, null is returned.
 */
fun <T : Comparable<T>> max(array: Iterable<T>?): T? = array?.maxOrNull()

/**
 * This method is like _.max except that it accepts iteratee which is invoked for each element in array to generate the criterion by which the value is ranked.
 */
fun <T, R : Comparable<R>> maxBy(array: Iterable<T>?, iteratee: (T) -> R): T? {
    return array?.maxByOrNull(iteratee)
}

/**
 * Computes the mean of the values in array.
 */
fun mean(array: Iterable<Number>?): Double {
    if (array == null) return Double.NaN
    val list = array.toList()
    if (list.isEmpty()) return Double.NaN
    return list.sumOf { it.toDouble() } / list.size
}

/**
 * This method is like _.mean except that it accepts iteratee which is invoked for each element in array to generate the criterion by which the value is averaged.
 */
fun <T> meanBy(array: Iterable<T>?, iteratee: (T) -> Number): Double {
    if (array == null) return Double.NaN
    val list = array.toList()
    if (list.isEmpty()) return Double.NaN
    return list.sumOf { iteratee(it).toDouble() } / list.size
}

/**
 * Computes the minimum value of array. If array is empty or falsey, null is returned.
 */
fun <T : Comparable<T>> min(array: Iterable<T>?): T? = array?.minOrNull()

/**
 * This method is like _.min except that it accepts iteratee which is invoked for each element in array to generate the criterion by which the value is ranked.
 */
fun <T, R : Comparable<R>> minBy(array: Iterable<T>?, iteratee: (T) -> R): T? {
    return array?.minByOrNull(iteratee)
}

/**
 * Multiply two numbers.
 */
fun multiply(multiplier: Double, multiplicand: Double): Double = multiplier * multiplicand
fun multiply(multiplier: Int, multiplicand: Int): Int = multiplier * multiplicand

/**
 * Computes number rounded to precision.
 */
fun round(number: Double, precision: Int = 0): Double {
    val factor = 10.0.pow(precision)
    return round(number * factor) / factor
}

/**
 * Subtract two numbers.
 */
fun subtract(minuend: Double, subtrahend: Double): Double = minuend - subtrahend
fun subtract(minuend: Int, subtrahend: Int): Int = minuend - subtrahend

/**
 * Computes the sum of the values in array.
 */
fun sum(array: Iterable<Number>?): Double = array?.sumOf { it.toDouble() } ?: 0.0

/**
 * This method is like _.sum except that it accepts iteratee which is invoked for each element in array to generate the criterion by which the value is summed.
 */
fun <T> sumBy(array: Iterable<T>?, iteratee: (T) -> Number): Double = array?.sumOf { iteratee(it).toDouble() } ?: 0.0

File: packages/modules/lodash/core/go/core/math/math.go
package math

import "math"

func Add(a float64, b float64) float64 {
	return a + b
}

func Ceil(number float64) float64 {
	return math.Ceil(number)
}

func Divide(a float64, b float64) float64 {
	return a / b
}

func Floor(number float64) float64 {
	return math.Floor(number)
}

func Max(array []float64) float64 {
	var maximum float64 = math.Inf(-8)
	for i := 0; i < len(array); i++ {
		var number float64 = array[i]
		if number > maximum {
			maximum = number
		}
	}
	return maximum
}

func Mean(array []float64) float64 {
	if len(array) == 0 {
		return 0
	}
	var _sum float64 = Sum(array)
	var mean float64 = _sum / float64(len(array))
	return mean
}

func Min(array []float64) float64 {
	var minimum float64 = math.Inf(8)
	for i := 0; i < len(array); i++ {
		var number float64 = array[i]
		if number < minimum {
			minimum = number
		}
	}
	return minimum
}

func Multiply(a float64, b float64) float64 {
	return a * b
}

func Round(number float64) float64 {
	return math.Round(number)
}

func Subtract(a float64, b float64) float64 {
	return a - b
}

func Sum(array []float64) float64 {
	var sum float64 = 0
	for i := 0; i < len(array); i++ {
		var number float64 = array[i]
		sum += number
	}
	return sum
}

func MaxBy(objects []map[string]float64, iteratee func(map[string]float64) float64) float64 {
	if len(objects) == 0 {
		return 0
	}
	var maximum float64 = math.Inf(-1)
	for _, obj := range objects {
		var val float64 = iteratee(obj)
		if val > maximum {
			maximum = val
		}
	}
	return maximum
}

func MeanBy(objects []map[string]float64, iteratee func(map[string]float64) float64) float64 {
	if len(objects) == 0 {
		return 0
	}
	var sum float64 = 0
	for _, obj := range objects {
		sum += iteratee(obj)
	}
	return sum / float64(len(objects))
}

func MinBy(objects []map[string]float64, iteratee func(map[string]float64) float64) float64 {
	if len(objects) == 0 {
		return 0
	}
	var minimum float64 = math.Inf(1)
	for _, obj := range objects {
		var val float64 = iteratee(obj)
		if val < minimum {
			minimum = val
		}
	}
	return minimum
}

func SumBy(objects []map[string]float64, iteratee func(map[string]float64) float64) float64 {
	var sum float64 = 0
	for _, obj := range objects {
		sum += iteratee(obj)
	}
	return sum
}
